## DSO104 - Data Wrangling and Visualization
### Lesson 2 - Data Transformations
______________
#### Page 22 - Hands-On

**Requirements**

You are working for Airbnb, and they are trying to improve their website. They've collected [data by unique id on gender, signup method, language, affiliates, devices, and browsers of website visitors](https://repo.exeterlms.com/documents/V2/DataScience/Data-Wrang-Visual/airbnb_test_users.zip).

They'd like to know the following:

* What is the total `signup_flow` for each device?

Make sure you use your newfound data aggregation skills to find the answer.

They would also like you to perform the following tasks:

* They need the `country` information from [this dataset](https://repo.exeterlms.com/documents/V2/DataScience/Data-Wrang-Visual/airbnb_sample_submission.zip) included in the `airbnb_test_users` file.
* Add additional users to the test file from [this dataset](https://repo.exeterlms.com/documents/V2/DataScience/Data-Wrang-Visual/airbnb_users.zip).
______________
**Import package and data**

In [1]:
import pandas as pd

In [29]:
airBnbTestUsers = pd.read_csv('/Users/hannah/Library/CloudStorage/GoogleDrive-gracesnouveaux@gmail.com/My Drive/Bethel Tech/Data Science/DSO104 Data Wrangling and Visualization/Lesson 2: Data Transformations/airbnb_test_users.csv')

**Preview data**

In [14]:
airBnbTestUsers

id date_account_created  timestamp_first_active  \
0      5uwns89zht             7/1/2014            2.014070e+13   
1      jtl0dijy2j             7/1/2014            2.014070e+13   
2      xx0ulgorjt             7/1/2014            2.014070e+13   
3      6c6puo6ix0             7/1/2014            2.014070e+13   
4      czqhjk3yfe             7/1/2014            2.014070e+13   
...           ...                  ...                     ...   
62091  cv0na2lf5a            9/30/2014            2.014090e+13   
62092  zp8xfonng8            9/30/2014            2.014090e+13   
62093  fa6260ziny            9/30/2014            2.014090e+13   
62094  87k0fy4ugm            9/30/2014            2.014090e+13   
62095  9uqfg8txu3            9/30/2014            2.014090e+13   

       date_first_booking     gender   age signup_method  signup_flow  \
0                     NaN     FEMALE  35.0      facebook            0   
1                     NaN  -unknown-   NaN         basic            0   
2                     NaN  -unknown-   NaN         basic            0   
3                     NaN  -unknown-   NaN         basic            0   
4                     NaN  -unknown-   NaN         basic            0   
...                   ...        ...   ...           ...          ...   
62091                 NaN  -unknown-  31.0         basic            0   
62092                 NaN  -unknown-   NaN         basic           23   
62093                 NaN  -unknown-   NaN         basic            0   
62094                 NaN  -unknown-   NaN         basic            0   
62095                 NaN     FEMALE  49.0         basic            0   

      language affiliate_channel affiliate_provider first_affiliate_tracked  \
0           en            direct             direct               untracked   
1           en            direct             direct               untracked   
2           en            direct             direct                  linked   
3           en            direct             direct                  linked   
4           en            direct             direct               untracked   
...        ...               ...                ...                     ...   
62091       en            direct             direct               untracked   
62092       ko            direct             direct               untracked   
62093       de            direct             direct                  linked   
62094       en         sem-brand             google                     omg   
62095       en             other              other           tracked-other   

      signup_app first_device_type  first_browser  
0          Moweb            iPhone  Mobile Safari  
1          Moweb            iPhone  Mobile Safari  
2            Web   Windows Desktop         Chrome  
3            Web   Windows Desktop             IE  
4            Web       Mac Desktop         Safari  
...          ...               ...            ...  
62091        Web   Windows Desktop             IE  
62092    Android     Android Phone      -unknown-  
62093        Web   Windows Desktop        Firefox  
62094        Web       Mac Desktop         Safari  
62095        Web   Windows Desktop         Chrome  

[62096 rows x 15 columns]

### What is the average `age` of those who use each web browser type?

In [44]:
airBnbTestUsers.groupby('first_browser')['age'].mean()

first_browser
-unknown-             41.526894
AOL Explorer          60.000000
Android Browser       42.822148
Apple Mail            26.750000
BlackBerry Browser    34.600000
Chrome                34.591678
Chrome Mobile         45.199058
Chromium              28.833333
CometBird             30.000000
Firefox               37.748143
IBrowse               51.000000
IE                    42.463107
IE Mobile             32.312500
IceWeasel             18.000000
Iron                  35.200000
Maxthon               31.000000
Mobile Firefox        31.800000
Mobile Safari         35.098486
Nintendo Browser            NaN
Opera                 38.444444
Opera Mini                  NaN
Opera Mobile          37.000000
Pale Moon             37.000000
Safari                36.416121
SeaMonkey                   NaN
Silk                  38.120000
SiteKiosk             57.000000
Sogou Explorer        28.000000
UC Browser            28.000000
Yandex.Browser        38.500000
wOSBrowser                

*Used `groupby` function to aggregate with a `mean` argument (lol ;) to find averages... I think it would be helpful for these to be sorted by the Age column (oldest at top) and to round the floats to convert them to integers (seems odd to have an average age of 45.199058), so I found options [here](https://sparkbyexamples.com/pandas/pandas-groupby-sort-within-groups/) and [here](https://datatofish.com/round-values-pandas-dataframe/)*

In [54]:
airBnbTestUsers.groupby('first_browser')['age'].mean().sort_values(ascending = False).round()

first_browser
AOL Explorer          60.0
SiteKiosk             57.0
IBrowse               51.0
Chrome Mobile         45.0
Android Browser       43.0
IE                    42.0
-unknown-             42.0
Yandex.Browser        38.0
Opera                 38.0
Silk                  38.0
Firefox               38.0
Opera Mobile          37.0
Pale Moon             37.0
Safari                36.0
Iron                  35.0
Mobile Safari         35.0
BlackBerry Browser    35.0
Chrome                35.0
IE Mobile             32.0
Mobile Firefox        32.0
Maxthon               31.0
CometBird             30.0
Chromium              29.0
Sogou Explorer        28.0
UC Browser            28.0
Apple Mail            27.0
IceWeasel             18.0
Nintendo Browser       NaN
Opera Mini             NaN
SeaMonkey              NaN
wOSBrowser             NaN
Name: age, dtype: float64

______________
### What is the total `signup_flow` for each device?

In [55]:
airBnbTestUsers.groupby('first_device_type')['signup_flow'].sum().sort_values(ascending = False)

first_device_type
iPhone                345343
Android Phone          90521
iPad                   28900
Android Tablet          5969
Mac Desktop             5736
Other/Unknown           4826
Windows Desktop         3866
Desktop (Other)           50
SmartPhone (Other)         0
Name: signup_flow, dtype: int64

*Used `groupby` function to aggregate with a `sum` argument to find totals and a `sort` argument to have highest totals at the top*
_________
### They need the `country` information from [this dataset](https://repo.exeterlms.com/documents/V2/DataScience/Data-Wrang-Visual/airbnb_sample_submission.zip) included in the `airbnb_test_users` file

**Import data**

In [35]:
airBnbSample = pd.read_csv('/Users/hannah/Library/CloudStorage/GoogleDrive-gracesnouveaux@gmail.com/My Drive/Bethel Tech/Data Science/DSO104 Data Wrangling and Visualization/Lesson 2: Data Transformations/airbnb_sample_submission.csv')

**Preview data**

In [32]:
airBnbSample

id country
0      5uwns89zht     NDF
1      jtl0dijy2j     NDF
2      xx0ulgorjt     NDF
3      6c6puo6ix0     NDF
4      czqhjk3yfe     NDF
...           ...     ...
62091  cv0na2lf5a     NDF
62092  zp8xfonng8     NDF
62093  fa6260ziny     NDF
62094  87k0fy4ugm     NDF
62095  9uqfg8txu3     NDF

[62096 rows x 2 columns]

*Because this dataset and the original both have an `id` column, I can merge using that column, without needing to explicitly define it – thanks, Python!*

**Merge datasets**

In [33]:
airBnbWithSample = pd.merge(airBnbTestUsers, airBnbSample)

**Preview data**

In [34]:
airBnbWithSample

id date_account_created  timestamp_first_active  \
0      5uwns89zht             7/1/2014            2.014070e+13   
1      jtl0dijy2j             7/1/2014            2.014070e+13   
2      xx0ulgorjt             7/1/2014            2.014070e+13   
3      6c6puo6ix0             7/1/2014            2.014070e+13   
4      czqhjk3yfe             7/1/2014            2.014070e+13   
...           ...                  ...                     ...   
62091  cv0na2lf5a            9/30/2014            2.014090e+13   
62092  zp8xfonng8            9/30/2014            2.014090e+13   
62093  fa6260ziny            9/30/2014            2.014090e+13   
62094  87k0fy4ugm            9/30/2014            2.014090e+13   
62095  9uqfg8txu3            9/30/2014            2.014090e+13   

       date_first_booking     gender   age signup_method  signup_flow  \
0                     NaN     FEMALE  35.0      facebook            0   
1                     NaN  -unknown-   NaN         basic            0   
2                     NaN  -unknown-   NaN         basic            0   
3                     NaN  -unknown-   NaN         basic            0   
4                     NaN  -unknown-   NaN         basic            0   
...                   ...        ...   ...           ...          ...   
62091                 NaN  -unknown-  31.0         basic            0   
62092                 NaN  -unknown-   NaN         basic           23   
62093                 NaN  -unknown-   NaN         basic            0   
62094                 NaN  -unknown-   NaN         basic            0   
62095                 NaN     FEMALE  49.0         basic            0   

      language affiliate_channel affiliate_provider first_affiliate_tracked  \
0           en            direct             direct               untracked   
1           en            direct             direct               untracked   
2           en            direct             direct                  linked   
3           en            direct             direct                  linked   
4           en            direct             direct               untracked   
...        ...               ...                ...                     ...   
62091       en            direct             direct               untracked   
62092       ko            direct             direct               untracked   
62093       de            direct             direct                  linked   
62094       en         sem-brand             google                     omg   
62095       en             other              other           tracked-other   

      signup_app first_device_type  first_browser country  
0          Moweb            iPhone  Mobile Safari     NDF  
1          Moweb            iPhone  Mobile Safari     NDF  
2            Web   Windows Desktop         Chrome     NDF  
3            Web   Windows Desktop             IE     NDF  
4            Web       Mac Desktop         Safari     NDF  
...          ...               ...            ...     ...  
62091        Web   Windows Desktop             IE     NDF  
62092    Android     Android Phone      -unknown-     NDF  
62093        Web   Windows Desktop        Firefox     NDF  
62094        Web       Mac Desktop         Safari     NDF  
62095        Web   Windows Desktop         Chrome     NDF  

[62096 rows x 16 columns]

*Merge looks like it was successful - same # of rows, with one more column at the end*
___________
### Add additional users to the test file from [this dataset](https://repo.exeterlms.com/documents/V2/DataScience/Data-Wrang-Visual/airbnb_users.zip)

**Import data**

In [56]:
airBnbUsers = pd.read_csv('/Users/hannah/Library/CloudStorage/GoogleDrive-gracesnouveaux@gmail.com/My Drive/Bethel Tech/Data Science/DSO104 Data Wrangling and Visualization/Lesson 2: Data Transformations/airbnb_users.csv')

**Preview data**

In [57]:
airBnbUsers

id date_account_created  timestamp_first_active  \
0       gxn3p5htnn            6/28/2010            2.009030e+13   
1       820tgsjxq7            5/25/2011            2.009050e+13   
2       4ft3gnwmtx            9/28/2010            2.009060e+13   
3       bjjt8pjhuk            12/5/2011            2.009100e+13   
4       87mebub9p4            9/14/2010            2.009120e+13   
...            ...                  ...                     ...   
213446  zxodksqpep            6/30/2014            2.014060e+13   
213447  mhewnxesx9            6/30/2014            2.014060e+13   
213448  6o3arsjbb4            6/30/2014            2.014060e+13   
213449  jh95kwisub            6/30/2014            2.014060e+13   
213450  nw9fwlyb5f            6/30/2014            2.014060e+13   

       date_first_booking     gender   age signup_method  signup_flow  \
0                     NaN  -unknown-   NaN      facebook            0   
1                     NaN       MALE  38.0      facebook            0   
2                8/2/2010     FEMALE  56.0         basic            3   
3                9/8/2012     FEMALE  42.0      facebook            0   
4               2/18/2010  -unknown-  41.0         basic            0   
...                   ...        ...   ...           ...          ...   
213446                NaN       MALE  32.0         basic            0   
213447                NaN  -unknown-   NaN         basic            0   
213448                NaN  -unknown-  32.0         basic            0   
213449                NaN  -unknown-   NaN         basic           25   
213450                NaN  -unknown-   NaN         basic           25   

       language affiliate_channel affiliate_provider first_affiliate_tracked  \
0            en            direct             direct               untracked   
1            en               seo             google               untracked   
2            en            direct             direct               untracked   
3            en            direct             direct               untracked   
4            en            direct             direct               untracked   
...         ...               ...                ...                     ...   
213446       en         sem-brand             google                     omg   
213447       en            direct             direct                  linked   
213448       en            direct             direct               untracked   
213449       en             other              other           tracked-other   
213450       en            direct             direct               untracked   

       signup_app first_device_type  first_browser country_destination  
0             Web       Mac Desktop         Chrome                 NDF  
1             Web       Mac Desktop         Chrome                 NDF  
2             Web   Windows Desktop             IE                  US  
3             Web       Mac Desktop        Firefox               other  
4             Web       Mac Desktop         Chrome                  US  
...           ...               ...            ...                 ...  
213446        Web       Mac Desktop         Safari                 NDF  
213447        Web   Windows Desktop         Chrome                 NDF  
213448        Web       Mac Desktop        Firefox                 NDF  
213449        iOS            iPhone  Mobile Safari                 NDF  
213450        iOS            iPhone      -unknown-                 NDF  

[213451 rows x 16 columns]

*Everything looks good other than the fact that the last column has a different name than the prior dataset; this one is more specific, so I'll rename the existing data to be able to then append this dataset to the original*

*Also, cool to see how quickly Python can handle so much data – I am used to Google Sheets which slow down noticeably when manipulating more than 1k rows at a time, and this is over 200k rows!*

**Rename `country` column in merged dataset to `country_destination`**

In [58]:
airBnbWithSample.rename(columns = {'country': 'country_destination'}, inplace = True)

airBnbWithSample

id date_account_created  timestamp_first_active  \
0      5uwns89zht             7/1/2014            2.014070e+13   
1      jtl0dijy2j             7/1/2014            2.014070e+13   
2      xx0ulgorjt             7/1/2014            2.014070e+13   
3      6c6puo6ix0             7/1/2014            2.014070e+13   
4      czqhjk3yfe             7/1/2014            2.014070e+13   
...           ...                  ...                     ...   
62091  cv0na2lf5a            9/30/2014            2.014090e+13   
62092  zp8xfonng8            9/30/2014            2.014090e+13   
62093  fa6260ziny            9/30/2014            2.014090e+13   
62094  87k0fy4ugm            9/30/2014            2.014090e+13   
62095  9uqfg8txu3            9/30/2014            2.014090e+13   

       date_first_booking     gender   age signup_method  signup_flow  \
0                     NaN     FEMALE  35.0      facebook            0   
1                     NaN  -unknown-   NaN         basic            0   
2                     NaN  -unknown-   NaN         basic            0   
3                     NaN  -unknown-   NaN         basic            0   
4                     NaN  -unknown-   NaN         basic            0   
...                   ...        ...   ...           ...          ...   
62091                 NaN  -unknown-  31.0         basic            0   
62092                 NaN  -unknown-   NaN         basic           23   
62093                 NaN  -unknown-   NaN         basic            0   
62094                 NaN  -unknown-   NaN         basic            0   
62095                 NaN     FEMALE  49.0         basic            0   

      language affiliate_channel affiliate_provider first_affiliate_tracked  \
0           en            direct             direct               untracked   
1           en            direct             direct               untracked   
2           en            direct             direct                  linked   
3           en            direct             direct                  linked   
4           en            direct             direct               untracked   
...        ...               ...                ...                     ...   
62091       en            direct             direct               untracked   
62092       ko            direct             direct               untracked   
62093       de            direct             direct                  linked   
62094       en         sem-brand             google                     omg   
62095       en             other              other           tracked-other   

      signup_app first_device_type  first_browser country_destination  
0          Moweb            iPhone  Mobile Safari                 NDF  
1          Moweb            iPhone  Mobile Safari                 NDF  
2            Web   Windows Desktop         Chrome                 NDF  
3            Web   Windows Desktop             IE                 NDF  
4            Web       Mac Desktop         Safari                 NDF  
...          ...               ...            ...                 ...  
62091        Web   Windows Desktop             IE                 NDF  
62092    Android     Android Phone      -unknown-                 NDF  
62093        Web   Windows Desktop        Firefox                 NDF  
62094        Web       Mac Desktop         Safari                 NDF  
62095        Web   Windows Desktop         Chrome                 NDF  

[62096 rows x 16 columns]

**Append final dataset to the merged dataset**

In [28]:
airBnbFinal = pd.concat([airBnbWithSample, airBnbUsers])

airBnbFinal

id date_account_created  timestamp_first_active  \
0       5uwns89zht             7/1/2014            2.014070e+13   
1       jtl0dijy2j             7/1/2014            2.014070e+13   
2       xx0ulgorjt             7/1/2014            2.014070e+13   
3       6c6puo6ix0             7/1/2014            2.014070e+13   
4       czqhjk3yfe             7/1/2014            2.014070e+13   
...            ...                  ...                     ...   
213446  zxodksqpep            6/30/2014            2.014060e+13   
213447  mhewnxesx9            6/30/2014            2.014060e+13   
213448  6o3arsjbb4            6/30/2014            2.014060e+13   
213449  jh95kwisub            6/30/2014            2.014060e+13   
213450  nw9fwlyb5f            6/30/2014            2.014060e+13   

       date_first_booking     gender   age signup_method  signup_flow  \
0                     NaN     FEMALE  35.0      facebook            0   
1                     NaN  -unknown-   NaN         basic            0   
2                     NaN  -unknown-   NaN         basic            0   
3                     NaN  -unknown-   NaN         basic            0   
4                     NaN  -unknown-   NaN         basic            0   
...                   ...        ...   ...           ...          ...   
213446                NaN       MALE  32.0         basic            0   
213447                NaN  -unknown-   NaN         basic            0   
213448                NaN  -unknown-  32.0         basic            0   
213449                NaN  -unknown-   NaN         basic           25   
213450                NaN  -unknown-   NaN         basic           25   

       language affiliate_channel affiliate_provider first_affiliate_tracked  \
0            en            direct             direct               untracked   
1            en            direct             direct               untracked   
2            en            direct             direct                  linked   
3            en            direct             direct                  linked   
4            en            direct             direct               untracked   
...         ...               ...                ...                     ...   
213446       en         sem-brand             google                     omg   
213447       en            direct             direct                  linked   
213448       en            direct             direct               untracked   
213449       en             other              other           tracked-other   
213450       en            direct             direct               untracked   

       signup_app first_device_type  first_browser country_destination  
0           Moweb            iPhone  Mobile Safari                 NDF  
1           Moweb            iPhone  Mobile Safari                 NDF  
2             Web   Windows Desktop         Chrome                 NDF  
3             Web   Windows Desktop             IE                 NDF  
4             Web       Mac Desktop         Safari                 NDF  
...           ...               ...            ...                 ...  
213446        Web       Mac Desktop         Safari                 NDF  
213447        Web   Windows Desktop         Chrome                 NDF  
213448        Web       Mac Desktop        Firefox                 NDF  
213449        iOS            iPhone  Mobile Safari                 NDF  
213450        iOS            iPhone      -unknown-                 NDF  

[275547 rows x 16 columns]

*Et voila!*